core/modules/loader/DoListItem.h

```diff
@@ -0,0 +1,180 @@
+/// few hours or even days).
+/// The DoListItems can cycle forever by just remaining on the
+/// DoList where it will run their actions when the timer run out,
+/// which is useful for monitoring status.
+/// Or they can be setup to run until they have completed once,
+/// a oneShot, which is useful for looking up or inserting keys.
+///
+/// A typical action would be sending out a UDP request for status
+/// every few seconds until a response is received. Then, after a
+/// few minutes with no updates, repeating that request to make sure
+/// the status hasn't changed.
+/// The system is supposed to notify others on changes, but these
+/// notifications can lost, so it makes sense to ask for one if
+/// nothing has been received for a while.
+class DoListItem : public std::enable_shared_from_this<DoListItem> {
+public:
+    using Ptr = std::shared_ptr<DoListItem>;
+
+    /// Children of this class *MUST* be created with shared pointers.
+    /// A factory function to enforce this is not practical since
+    /// this class is meant to serve as a base class for unknown
+    /// future purposes. Sadly, the compiler doesn't enforce the rule.
+    DoListItem() = default;
+
+    DoListItem(DoListItem const&) = delete;
+    DoListItem& operator=(DoListItem const&) = delete;
+
+    virtual ~DoListItem() = default;
+
+    util::CommandTracked::Ptr runIfNeeded(TimeOut::TimePoint now) {
+        std::lock_guard<std::mutex> lock(_mtx);
+        if (_command == nullptr) {
+            if (_isOneShotDone()) { return nullptr; }
+            if ((_needInfo || _timeOut.due(now)) && _timeRequest.due(now)) {
+                _timeRequest.triggered();
+                    _command = createCommand();
+                return _command;
+            }
+        } else if (_command->isFinished()) {
+            _command.reset(); // Allow the command to be sent again later.
+        }
+        return nullptr;
+    }
+
+    bool isAlreadyOnList() { return _addedToList; }
+
+    /// Returns original value of _addedToList.
+    bool setAddedToList(bool value) {
+        return _addedToList.exchange(value);
+    }
+
+    /// @return true if this item should be removed from the list.
+    bool shouldRemoveFromList() {
+        std::lock_guard<std::mutex> lock(_mtx);
+        return (_isOneShotDone() || _remove);
+    }
+
+    /// The info has been updated, so no need to ask for it for a while.
+    void infoReceived() {
+        std::lock_guard<std::mutex> lock(_mtx);
+        _needInfo = false;
+        _timeOut.triggered();
+    }
+
+    void setNeedInfo() {
+        std::lock_guard<std::mutex> lock(_mtx);
+        _needInfo = true;
+    }
+
+    DoListItem::Ptr getDoListItemPtr() {
+        return shared_from_this();
+    }
+
+    void setTimeOut(std::chrono::milliseconds timeOut) { _timeOut.setTimeOut(timeOut); }
+
+    virtual util::CommandTracked::Ptr createCommand()=0;
+
+    /// The only class that needs access to most of the is DoList.
+    friend class DoList;
```

Comment:
this should be done with accessors, especially to guarantee the locking requirements.

Reply:
Deleted.